samcli/commands/pipeline/init/interactive_init_flow.py

```diff
@@ -0,0 +1,244 @@
+"""
+Interactive flow that prompts that users for pipeline template (cookiecutter template) and used it to generate
+pipeline configuration file
+"""
+import logging
+import os
+import shutil
+from pathlib import Path
+from typing import Dict, List
+
+import click
+
+from samcli.cli.main import global_cfg
+from samcli.commands.exceptions import PipelineTemplateCloneException
+from samcli.lib.cookiecutter.interactive_flow import InteractiveFlow
+from samcli.lib.cookiecutter.interactive_flow_creator import InteractiveFlowCreator
+from samcli.lib.cookiecutter.template import Template
+from samcli.lib.utils.git_repo import GitRepo, CloneRepoException
+from samcli.lib.utils.osutils import rmtree_callback
+from .pipeline_templates_manifest import PipelineTemplateManifest, PipelineTemplatesManifest
+
+LOG = logging.getLogger(__name__)
+shared_path: Path = global_cfg.config_dir
+APP_PIPELINE_TEMPLATES_REPO_URL = "https://github.com/aws/aws-sam-cli-pipeline-init-templates.git"
+APP_PIPELINE_TEMPLATES_REPO_LOCAL_NAME = "aws-sam-cli-app-pipeline-templates"
+CUSTOM_PIPELINE_TEMPLATE_REPO_LOCAL_NAME = "custom-pipeline-template"
+
+
+def do_interactive() -> None:
+    """
+    An interactive flow that prompts the user for pipeline template (cookiecutter template) location, downloads it,
+    runs its specific questionnaire then generates the pipeline config file based on the template and user's responses
+    """
+    click.echo("Which pipeline template source would you like to use?")
+    click.echo("\t1 - AWS Quick Start Pipeline Templates\n\t2 - Custom Pipeline Template Location")
+    location_choice = click.prompt("Choice", type=click.Choice(["1", "2"]), show_choices=False)
+    if location_choice == "2":
+        _generate_from_custom_location()
+    else:
+        _generate_from_app_pipeline_templates()
```

Comment:
Can we abstract this away? Showing a bunch of choices and calling corresponding callbacks is generic and well-defined. I noticed this is also being used in #2811, but it's fragile (need to ensure the choices and prompts match) and hard to read (coupled stuff like the choice value, expected response and callback are tightly coupled but scattered).

Reply:
do you mean not depending on the fact that click is guaranteed to return either "1" or "2" in this case? if so, yes we can but this just adds one more unnecessarily branch.

Reply:
I mean it's too verbose and error-prone.

We can observe that the multiple-choice prompts always (1) print the choices in a specific way (ordered list with tabs and `-`) (2) using specific inputs (a numbered list). Having to read and write this every time and ensuring it's consistent is too much cognitive complexity.

Instead we can refactor into something like this:

```python
def prompt_choices(choices: List[str]) -> int:
    """Prompt choice and return its index."""
    for i, choice in enumerate(choices):
        click.echo(f"\t{i + 1} - {choice}")
    inputs = [str(i + 1) for i in range(len(choices))]
    input = click.prompt("Choice", type=click.Choice(inputs), show_choices=False)
    return int(input) - 1
```

This way,

```python
    click.echo("\t1 - AWS Quick Start Pipeline Templates\n\t2 - Custom Pipeline Template Location")
    location_choice = click.prompt("Choice", type=click.Choice(["1", "2"]), show_choices=False)
    if location_choice == "2":
        _generate_from_custom_location()
    else:
        _generate_from_app_pipeline_templates()
```
 
becomes

```python
    location_choice = prompt_choice([
        "AWS Quick Start Pipeline Templates",
        "Custom Pipeline Template Location",
    ])
    if location_choice == 0:
        _generate_from_app_pipeline_templates()
    else:
        _generate_from_custom_location()
```

And this (from below)

```python
choices = list(map(str, range(1, len(available_pipeline_templates_manifests) + 1)))
for index, template in enumerate(available_pipeline_templates_manifests):
    click.echo(f"\t{index + 1} - {template.name}")
choice = click.prompt(text="Choice", show_choices=False, type=click.Choice(choices))
```

becomes

```python
choice = prompt_choice(available_pipeline_templates_manifests)
```